### ISSUE TITLE
Use Algolia DocSearch for search

### ISSUE THREAD

**Issue by ghys** [openhab/website#73]
The lack of full-text search has been brought up several times, for instance in #28, and is a valid point.

VuePress has built-in support for [Algolia DocSearch](https://community.algolia.com/docsearch/) which offers that and presumably better search results, for free. It's already used in a number of high-profile websites.
See it in action of a similar VuePress site here: https://www.vuemeetups.org

Using it instead of the default header-based search is easy and it makes sense to consider it.
1. A request should be submitted at https://community.algolia.com/docsearch/ (by @kaikreuzer as the owner of the website);
The URLs to crawl should be both https://www.openhab.org/docs/ + https://www.openhab.org/addons/
2. They will add our site to their crawler here: https://github.com/algolia/docsearch-configs/tree/master/configs and reply with configuration information by email;
3. I will adapt the config as documented here: https://vuepress.vuejs.org/default-theme-config/#algolia-search;
3. @MHerbst: a mention in the Privacy Policy would have to be added I think, since their API would be called when searching? Their privacy policy is here: https://www.algolia.com/policies/privacy - they appear to have updated it for the GDPR.

What do you think, should we at least give it a try?

**Comment by MHerbst** [comment]
Looking good. I would give it a try. Can the API calls embedded in a way that the Agolia Server is only contacted if the user executes a search? This means that no JavaScript should be loaded from Algolia's server before the user executes a search.  This would reduce problems with the GDPR because the user has the choice whether he wants to contact Algolia's server or not.

**Comment by ghys** [comment]
@MHerbst good question - I took a quick look how it's behaving on https://www.vuemeetups.org

Apparently it adds a NPM dependency to docsearch.js (https://github.com/algolia/docsearch) - which means it would be served from openhab.org (https://www.openhab.org/assets/js/vendors~docsearch.(hash).js)  instead of from a CDN.
As far as I can tell, it also doesn't make external requests unless you actually start searching, and doesn't set any cookies.

**Comment by MHerbst** [comment]
@ghy It's good that the JavaScript file is not loaded from Algolia's server. Therefore I would assume that the server is only contacted if a search is executed. We can verify this with Browser console.

**Comment by ghys** [comment]
Yes, I already confirmed that:
> As far as I can tell, it also doesn't make external requests unless you actually start searching, and doesn't set any cookies.

**Comment by ghys** [comment]
Also note they have a [Discourse plugin](https://github.com/algolia/discourse-algolia) which looks interesting too, and could be used to have the same search experience for the docs and the forums - I could probably put a similar-looking search box in the header. It looks a little more involved though (this needs an account with them), so I would not consider it now.

**Comment by kaikreuzer** [comment]
Hey @Shipow, great, thanks! I can fill out the form asap - but as mentioned by @ghys [above](https://github.com/openhab/website/issues/73#issue-331667858), we actually need two urls to be crawled - how do I best achieve this? Shall I submit the form twice for both urls (with the same e-mail address)?

**Comment by Shipow** [comment]
Hello @kaikreuzer, yes you can submit it only once, it's on the same domain, we'll work on the setup together. If you have some resources available, making a dedicated search UI for the addons would be better. I did an example for Zapier here https://codesandbox.io/s/v67vlxjkw0. I think it's interesting because then you can search in action and triggers supported.

**Comment by kaikreuzer** [comment]
Ok, I have just submitted the form!

**Comment by s-pace** [comment]
👋 all,

Thank you for these relevant details!

I've successfully configured the underlying crawler and it will now run every 24h. 

You're now a few steps away from having it working on your website: 
- Copy the following CSS/JS snippets and add them to your page 

```html
<!-- at the end of the HEAD --> 
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" /> 

<!-- at the end of the BODY --> 
<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script> 
<script type="text/javascript"> docsearch({ 
 apiKey: 'af17a113c6a11af8057592a3120ffd3b', 
 indexName: 'openhab', 
 inputSelector: '### REPLACE ME ####', 
 algoliaOptions: { 'facetFilters': ["tags:$TAGS"] }, 
 debug: false // Set debug to true if you want to inspect the dropdown 
}); 
</script> 
``
- Add a search input in your page if you don't have any yet. Then update the inputSelector value in JS snippet to a CSS selector that targets your search input field. 

- Replace `$TAGS` with the tags you want to search on. The list of possible tags is hardcoded in the config. So as of today you have: `addons`, `docs` 
   For example if you want to refine the search to the tags "addons" just specify:
```js
 'facetFilters': ["tags:addons"] 
```

- Optionally customize the look and feel by following [the DocSearch documentation](https://community.algolia.com/docsearch/documentation/). 
- You can also check your configuration in [our github repo](https://github.com/algolia/docsearch-configs/blob/master/configs/openhab.json). 

I have granted @kaikreuze access to the free analytics we do provide for your index. Check your mailbox. Feel free to send me email addresses of people to add thanks to this private channel.

Please open a pull request if want to leverage your configuration! 

We'd also be happy to get your feedback and thoughts about DocSearch - so we can continue to improve it.

**Comment by ghys** [comment]
It's live now at https://www.openhab.org/ and seems to work pretty well with the defaults.

Thanks!

**Comment by s-pace** [comment]
It is because it was customised as much as possible :)

🎉 news! 🙏

**Comment by ghys** [comment]
@MHerbst: could you check whether it is appropriate to mention Algolia at https://www.openhab.org/privacy.html#plugins-and-tools (and also remove the "Google Web Fonts" paragraph because the fonts are now served locally)?

**Comment by Shipow** [comment]
Thanks @s-pace for your reactivity.
Dear OpenHAB team, let's stay in touch to follow up on your integration and see what users think.
Also our community forum is a good place to get some tricks or ask questions.
And keep the good work on!

**Comment by ghys** [comment]
Love the fuzzy string matching!

![image](https://user-images.githubusercontent.com/2004147/43164290-1585cecc-8f91-11e8-989f-582d57b0605e.png)


@s-pace @Shipow I noticed some pages are duplicated in the index, for example "openHABian" and "sitemaps".
If you try searching for e.g. "raspberry" or "chart", you will get duplicate results, one of which leading to an invalid page (https://www.openhab.org/docs/configuration/sitemaps/#element-type-chart), the second one being correct (https://www.openhab.org/docs/configuration/sitemaps.html#element-types).

I understand there could be invalid links and the site lacks a sitemap which isn't supported by the web framework we're using, but can this be solved with a change to the indexing configuration?

**Comment by s-pace** [comment]
Let me check @ghys

**Comment by s-pace** [comment]
Fixing by https://github.com/algolia/docsearch-configs/commit/8e19c724ce2624b09750e4019c7ae2cab6358406

You have wrong URL still available. `stop_urls` solved it.

Happy to have your detail feedback as @Shipow said

**Comment by ghys** [comment]
Outstanding, thank you @s-pace for your quick support!

**Comment by Shipow** [comment]
We don't have yet a dedicated newsletter for Docsearch yet but the news are relayed through Algolia's newsletter, you can add yourself by subscribing on the blog or create an Algolia account and opt-in for the coms. Actually you might need an account anyway to access the search analytics.
We'll get a new website for Docsearch soon. Cheers.

**Comment by s-pace** [comment]
Indeed. You can also follow [our dedicated documentation](https://community.algolia.com/docsearch/)

And follow/star our two open source repo:
[- Front search-UI](https://github.com/algolia/docsearch)
[- Crawler](https://github.com/algolia/docsearch-scraper)

**Comment by kaikreuzer** [comment]
> @kaikreuzer did you sign Algolia's DPA (https://www.algolia.com/pdf/DPA-latest.pdf)?

@MHerbst Not yet, will do so the next days!
